src/components/network/network-loader.js

```diff
@@ -19,117 +19,286 @@ genotet.utils.inherit(genotet.NetworkLoader, genotet.ViewLoader);
- * @param {string} geneRegex Regex that selects the genes to be updated.
+ * @param {string} inputGene Genes that selects the genes to be updated.
+ * @param {boolean} isRegex Update is based on regex or not.
  */
-genotet.NetworkLoader.prototype.updateGenes = function(method, geneRegex) {
-  var regex = this.data.geneRegex;
+genotet.NetworkLoader.prototype.updateGenes = function(method, inputGene,
+                                                       isRegex) {
+  var genes = this.prepareGene_(inputGene, isRegex);
+  if (genes.length == 0) {
+    genotet.warning('no gene input found');
+    return;
+  }
+
   switch (method) {
     case 'set':
-      // Totally replace the regex.
-      regex = geneRegex;
+      this.loadNetwork_(this.data.network.fileName, genes);
       break;
     case 'add':
-      // Concat the two regex's. We need to include the previously existing
-      // genes too so as to find the edges between the new genes and old
-      // ones.
-      regex += '|' + geneRegex;
+      this.addGene_(genes);
       break;
     case 'remove':
-      // Remove genes do not need communication with the server.
-      this.removeGenes_(geneRegex);
-      // Return immediately. No ajax.
-      return;
+      this.deleteGene_(genes);
+      break;
   }
-  this.load(this.data.fileName, regex);
 };
 
+/**
+ * Fetches the incident edges of a given node.
+ * @param {!Object} node Node of which the incident edges are queried.
+ */
+genotet.NetworkLoader.prototype.incidentEdges = function(node) {
+  var params = {
+    type: 'incident-edges',
+    fileName: this.data.network.fileName,
+    gene: node.id
+  };
+  this.get(genotet.data.serverURL, params, function(data) {
+    this.data.incidentEdges = data;
+    this.signal('incidentEdges');
+  }.bind(this), 'cannot get incident edges');
+};
 
 /**
- * Removes the selected genes from the current network data.
- * @param {string} geneRegex Regex selecting the genes to be removed.
+ * Adds genes to the network, and the edges to the original graph.
+ * @param {!Array<string>} genes Names of the genes.
  * @private
  */
-genotet.NetworkLoader.prototype.removeGenes_ = function(geneRegex) {
-  var regex;
-  try {
-    regex = RegExp(geneRegex, 'i');
-  } catch (e) {
-    genotet.error('invalid gene regex', geneRegex);
-    return;
-  }
-  this.data.nodes = _.filter(this.data.nodes, function(node) {
-    return !node.id.match(regex);
+genotet.NetworkLoader.prototype.addGene_ = function(genes) {
+  var oldGenes = {};
+  this.data.network.nodes.forEach(function(node) {
+    oldGenes[node.id] = true;
   });
-  this.data.edges = _.filter(this.data.edges, function(edge) {
-    return !edge.source.match(regex) && !edge.target.match(regex);
+  var newGenes = [];
+  genes.forEach(function(gene) {
+    if (!(gene in oldGenes)) {
+      newGenes.push(gene);
+    }
   });
 
-  this.updateGeneRegex_();
-  this.signal('geneRemove');
+  if (newGenes.length == 0) {
+    genotet.warning('Genes are already in graph');
```

Comment:
'genes are already in the network'. Lowercase the first word and avoid using CS terms for user messages.